misk/src/main/kotlin/misk/web/interceptors/ConcurrencyLimitsInterceptor.kt

```diff
@@ -0,0 +1,79 @@
+package misk.web.interceptors
+
+import com.netflix.concurrency.limits.Limiter
+import com.netflix.concurrency.limits.limiter.SimpleLimiter
+import misk.Action
+import misk.exceptions.StatusCode
+import misk.logging.getLogger
+import misk.web.AvailableWhenDegraded
+import misk.web.ConcurrencyLimitsOptIn
+import misk.web.NetworkChain
+import misk.web.NetworkInterceptor
+import java.time.Clock
+import javax.inject.Inject
+import javax.inject.Singleton
+import kotlin.reflect.full.findAnnotation
+
+/**
+ * Detects degraded behavior and sheds requests accordingly. Internally this uses adaptive limiting
+ * as implemented by Netflix's [concurrency-limits][concurrency_limits] library. See
+ * [AvailableWhenDegraded] for further documentation.
+ *
+ * [concurrency_limits]: https://github.com/Netflix/concurrency-limits/
+ */
+internal class ConcurrencyLimitsInterceptor internal constructor(
+  private val actionName: String,
+  private val limiter: Limiter<String>
+) : NetworkInterceptor {
+
+  override fun intercept(chain: NetworkChain) {
+    val listener: Limiter.Listener? = limiter.acquire(actionName).orElse(null)
+
+    if (listener == null) {
+      logger.error { "concurrency limits interceptor shedding $actionName" }
```

Comment:
we probably don't want log everytime we shed. ideally we are incrementing a metric instead

Reply:
Will do, for follow-up though.